### PR TITLE
8292695: SIGQUIT and jcmd attaching mechanism does not work with signal chaining library

### PR DESCRIPTION
**Pull request by caoman**
Hi all,

Could anyone review this bug fix? See https://bugs.openjdk.org/browse/JDK-8292695 for details.

I changed the temporary handler for SIGQUIT to use a dummy function, and use `os::signal()` to set it up, just as `os::initialize_jdk_signal_support()` does.
It is possible that just moving the `set_signal_handler(BREAK_SIGNAL, false);` in `install_signal_handlers()` outside of the window bounded by `JVM_{begin|end}_signal_setting()` could also fix this bug. However, `set_signal_handler()` and `JVM_HANDLE_XXX_SIGNAL()` are currently used for signals that support chaining and periodically check, which do not apply to SIGQUIT. I think it is cleaner to use different functions for SIGQUIT.

I also added a test to check that sending SIGQUIT should produce a thread dump on stdout, with and without using libjsig.so.

-Man

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292695](https://bugs.openjdk.org/browse/JDK-8292695): SIGQUIT and jcmd attaching mechanism does not work with signal chaining library

### Reviewers
 * [Xin Liu](https://openjdk.org/census#xliu) (@navyxliu - Committer)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/9955/head:pull/9955` \
`$ git checkout pull/9955`

Update a local copy of the PR: \
`$ git checkout pull/9955` \
`$ git pull https://git.openjdk.org/jdk pull/9955/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 9955`

View PR using the GUI difftool: \
`$ git pr show -t 9955`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/9955.diff">https://git.openjdk.org/jdk/pull/9955.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/os/posix/signals_posix.cpp [hunk 1]
```diff
@@ -1339,6 +1333,21 @@ void install_signal_handlers() {
...
+    // prematurely. Late initialization will overwrite BREAK_SIGNAL's handler to
+    // UserHandler.
+    // Note that HotSpot does NOT support signal chaining for BREAK_SIGNAL.
+    // We have to set it outside the window bounded by libjsig's
+    // JVM_begin_signal_setting and JVM_end_signal_setting above, because the
+    // window is intended for signals that support chaining. Otherwise libjsig
+    // would prevent us from overwriting BREAK_SIGNAL's handler to UserHandler.
+    // We also use os::signal() and a dummy handler to avoid special-casing
+    // set_signal_handler() and JVM_HANDLE_XXX_SIGNAL().
+    os::signal(BREAK_SIGNAL, CAST_FROM_FN_PTR(void*, DummySIGBREAKHandler));
```
Comment (navyxliu) [inline src/hotspot/os/posix/signals_posix.cpp:1348]: Bummer. This is what I tried to avoid. Sorry!

By moving sigaction of BREAK_SIGNAL out of the "window" -- [begin_signal_setting, end_signal_setting], `libjsig` will not record the old action of 'BREAK_SIGNAL', so HotSpot still has freedom to replace it with **UserHandler**, which can handle ThreadDump request.

I agree that your approach is less intrusive and is better than previous one!

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+ */
+
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Map;
+import jdk.test.lib.Platform;
+import jdk.test.lib.Utils;
+import jdk.test.lib.process.ProcessTools;
+import jdk.test.lib.process.OutputAnalyzer;
+import vm.share.ProcessUtils;
```
[the remark this answers is not in the document]
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: There are a few serviceability dcmd jtreg tests that already depend on /vmTestbase:
- test/hotspot/jtreg/serviceability/dcmd/framework/HelpTest.java
- test/hotspot/jtreg/serviceability/dcmd/framework/VMVersionTest.java
- test/hotspot/jtreg/serviceability/dcmd/framework/Invalidcommandtest.java

Is there a real downside to depend on /vmTestbase?

Agree that it is a good suggestion to add a parameter to ProcessTools.sendCtrlBreak() and move it to test/lib. This function is only used in one other place datadumpreq001.java, so it's probably easy to change.
However, it's probably better to make a separate test-only RFE to move this function and add a test for JDK-8279124 together.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+
+public class TestBreakSignalThreadDump {
+
+    static class TestProcess {
+	static {
```
[the remark this answers is not in the document]
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java]: Sorry, I used wrong coding style for formatter. Fixed.

**Comment by tstuefe** [comment]
@caoman Im interested in what your launcher does, or is, since I have almost never encountered signal chaining in the field. Would you mind sharing what product the launcher belongs to, and why you enable signal chaining?

**Review by navyxliu — APPROVED** [review]
LGTM. I am not a reviewer.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (navyxliu) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: make sense.

[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: There can be issues using vmTestbase as a library, simply due to the fact that jtreg does not handle libraries particularly well and we can end up with NoClassDefFoundError at runtime due to "split packages".

**Review by dholmes-ora — COMMENTED** [review]
Still trying to gauge the cause and effects here ...

In the meantime, should the test not be put under runtime/jsig ?

**Comment by caoman** [comment]
@tstuefe Our launcher is equipped with almost all our Java applications. It basically does what bin/java does, but with several additional features. One main motivation to have our own launcher, is to static link all user-defined JNI dependency libraries with the launcher, so the applications do not need to use `System.loadLibrary()` to load `.so` files dynamically. Using static linking instead of dynamically loading `.so` files is core characteristic of our deployment environment for almost all languages (C, C++, Java, Python, Go). @jianglizhou is actually working on a project that also statically links HotSpot (`libjvm.so` and other `.so` files) with the our launcher as well.

Regarding to the signal chaining support from our launcher, we found some existing applications that `LD_PRELOAD`s `libjsig.so`. We'd like to support it in our launcher so they don't need to depend on `libjsig.so`. Those applications need signal chaining, because they use an internal C++ library that could install custom signal handlers for error handling.

**Comment by caoman** [comment]
> Still trying to gauge the cause and effects here ...
> 
> In the meantime, should the test not be put under runtime/jsig ?

The test tests SIGQUIT behavior with and without libjsig.so. So it does not necessarily depend on libjsig.so. I could move it if you prefer runtime/jsig.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: Should I implement sendCtrlBreak() outside of /vmTestbase in this PR then?
The new test did run and pass on Github pre-submit check.

[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: I'm not sure we need yet-another-signalling-mechanism for tests. Did you look at  ./hotspot/jtreg/runtime/signal/ ? I'm unclear why TestSigquit is not failing due to this issue?

[the remark this answers is not in the document]
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: Yes, I saw the tests under hotspot/jtreg/runtime/signal. TestSigquit is not failing due to the `-Xrs` flag, which prevents the JVM from installing any signal handler for SIGQUIT, so that the test can still install its own handler.

I didn't find any test to check that SIGQUIT would print a thread dump on stdout. It is good to have a test for this functionality.

[the remark this answers is not in the document]
Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:51]: Okay stick with using the vmTestbase utilities then; if they cause a problem we can address it then. The split library problem tends to show up only on very specific testing combinations.

**Comment by tstuefe** [comment]
> @tstuefe Our launcher is equipped with almost all our Java applications. It basically does what bin/java does, but with several additional features. One main motivation to have our own launcher, is to static link all user-defined JNI dependency libraries with the launcher, so the applications do not need to use `System.loadLibrary()` to load `.so` files dynamically. Using static linking instead of dynamically loading `.so` files is core characteristic of our deployment environment for almost all languages (C, C++, Java, Python, Go). @jianglizhou is actually working on a project that also statically links HotSpot (`libjvm.so` and other `.so` files) with the our launcher as well.

That is interesting. Do you link statically to improve startup speed, or for some other reason?

**Comment by caoman** [comment]
> > @tstuefe Our launcher is equipped with almost all our Java applications. It basically does what bin/java does, but with several additional features. One main motivation to have our own launcher, is to static link all user-defined JNI dependency libraries with the launcher, so the applications do not need to use `System.loadLibrary()` to load `.so` files dynamically. Using static linking instead of dynamically loading `.so` files is core characteristic of our deployment environment for almost all languages (C, C++, Java, Python, Go). @jianglizhou is actually working on a project that also statically links HotSpot (`libjvm.so` and other `.so` files) with the our launcher as well.
> 
> That is interesting. Do you link statically to improve startup speed, or for some other reason?

It is not about startup speed, but mainly to ensure correct version and simplify deployment. It is very similar to the advantages of static library as stated on Wikipedia: https://en.wikipedia.org/wiki/Static_library.
@jianglizhou will share more details about this project in a separate email thread or in a public talk soon, so stay tuned.

**Comment by caoman** [comment]
@dholmes-ora Let me know if there's any other issue, or if I could help explain the cause and effects better.

File: src/hotspot/os/posix/signals_posix.cpp [hunk 4]
```diff
@@ -1339,6 +1333,21 @@ void install_signal_handlers() {
...
+  if (!ReduceSignalUsage) {
+    // This is just for early initialization phase. Intercepting the signal here
+    // reduces the risk that an attach client accidentally forces HotSpot to quit
+    // prematurely. Late initialization will overwrite BREAK_SIGNAL's handler to
+    // UserHandler.
+    // Note that HotSpot does NOT support signal chaining for BREAK_SIGNAL.
+    // We have to set it outside the window bounded by libjsig's
+    // JVM_begin_signal_setting and JVM_end_signal_setting above, because the
+    // window is intended for signals that support chaining. Otherwise libjsig
+    // would prevent us from overwriting BREAK_SIGNAL's handler to UserHandler.
```
Comment (dholmes-ora) [inline src/hotspot/os/posix/signals_posix.cpp:1345]: Okay but doesn't that also mean that libjsig will now allow application code to install its own BREAK_SIGNAL handler, when it should be prevented from doing so???

Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp:1345]: Yes, this seems always the case regardless of this change. The application should use `-Xrs`/`-XX:+ReduceSignalUsage` if it needs to install its own handler for BREAK_SIGNAL.

This is also the reason why TestSigquit.java sets `-Xrs`, see test/hotspot/jtreg/runtime/signal/README:
> SIGQUIT, SIGTERM, SIGINT, and SIGHUP signals cannot be chained.
> If the application needs to handle these signals, the -Xrs option needs
> to be specified. So, test these signals only with -Xrs flag.

Comment (dholmes-ora) [reply src/hotspot/os/posix/signals_posix.cpp:1345]: My point is that this change seems to allow installation of a BREAK_SIGNAL handler even without -Xrs!

Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp:1345]: > My point is that this change seems to allow installation of a BREAK_SIGNAL handler even without -Xrs!

Got it. This is indeed the case. However, prior to [JDK-8279124](https://bugs.openjdk.org/browse/JDK-8279124), HotSpot also behaved in the same way, that it allows installation of a custom BREAK_SIGNAL handler even without -Xrs.

I verified with a modified TestSigquit.java, [here](https://github.com/caoman/jdk/commit/346132aa5af8d59406d0b98567d18d16868a3a32). It tests the "postpre" and "postpost" mode without -Xrs, which both uses libjsig.so:
- Without any change, the test times out waiting for the custom signal handler to be called.
- With JDK-8279124 [backed out](https://github.com/caoman/jdk/commit/0983baad17d78cf64e1f604abb232e581700a662), OR with change in this PR, the test passes.

Since HotSpot does not support chaining for SIGQUIT, I think it is OK to leave the behavior undefined if user installs a custom SIGQUIT handler without -Xrs. I.e. the custom handler may or may not work, and it could break HotSpot functionality. libjsig is not designed to prevent installing custom SIGQUIT handler.

Comment (dholmes-ora) [reply src/hotspot/os/posix/signals_posix.cpp:1345]: I am somewhat surprised by this. My reading of the jsig code is that once the VM has installed a SIGQUIT handler during VM init, then subsequent attempts to install a user handler will add the handler to the set to be chained, but because this is a VM signal the VM will never call that chained handler. With the proposed changes if the SIGQUIT handler has not been installed by the VM then the user handler will be installed directly - though I think the VM's handler would then replace that.

I find this all rather hard to follow and evaluate. As this seems to fix an issue I guess we can take it as-is and then keep an eye out for any unintended consequences. Though they will likely only come from external users so not something we will know till after the next release.

**Comment by dholmes-ora** [comment]
@tstuefe are you able to look at this one? Thanks.

**Comment by caoman** [comment]
@tstuefe Any feedback? I'd like to get this submitted soon to unblock our internal use cases relying on SIGQUIT and jcmd.

**Comment by tstuefe** [comment]
Sorry, missed this one. I'll take a look later today.

**Comment by tstuefe** [comment]
@caoman While looking at this, I think I found an unrelated bug in libjsig. https://bugs.openjdk.org/browse/JDK-8293466 Could you eyeball it and tell me if my assumption is correct?

**Comment by caoman** [comment]
Thanks for filing the other bug. Responded in that bug.

**Comment by tstuefe** [comment]
Looking at this, this gets more and more interesting. One thing, libjsig does nothing to prevent user code from changing thread/process sigmask. @caoman has that ever been a problem at Google?

**Comment by tstuefe** [comment]
Found another minor one: https://bugs.openjdk.org/browse/JDK-8293494.

I'm looking at the change now, though.

**Comment by caoman** [comment]
> One thing, libjsig does nothing to prevent user code from changing thread/process sigmask. @caoman has that never been a problem at Google?

We don't actually use libjsig directly. Our launcher reimplemented signal chaining support using the same JVM_begin_signal_setting/JVM_end_signal_setting/JVM_get_signal_action interface.

Our launcher's implementation has calls to pthread_sigmask, but the goal seems to make the overridden sigaction() async-signal-safe. It doesn't prevent user code from changing thread/process sigmask either. However, I haven't heard of issues with user code changing sigmask.

**Comment by caoman** [comment]
> But there is nothing that prevents us from moving the SIGQUIT handler installation up in time to the hotspot signal handler initialization. Nobody says the SIGQUIT handler has to be installed after the signal dispatcher thread started.

Thank you for the great idea and experiment! Let me try this approach and rerun the tests.

File: src/hotspot/os/posix/signals_posix.cpp (hunk 4 above)
Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp]: Updated with tstuefe@'s recommendations. It still keeps the behavior that a user can install a SIGQUIT handler without -Xrs.

**Comment by tstuefe** [comment]
> > > But there is nothing that prevents us from moving the SIGQUIT handler installation up in time to the hotspot signal handler initialization. Nobody says the SIGQUIT handler has to be installed after the signal dispatcher thread started.
> > 
> > 
> > Thank you for the great idea and experiment! Let me try this approach and rerun the tests.
> 
> Switched to the approach that installs the handler only once.
> 
> The SIGQUIT handler has to be installed after jdk_misc_signal_init() has initialized `pending_signals` and `sig_semaphore`, which are used by `os::signal_notify()`. This means SIGQUIT handler is still installed outside libjsig's JVM_begin_signal_setting/JVM_end_signal_setting window, so it not tracked by libjsig like other HotSpot signals.

You can just move `jdk_misc_signal_init` up a notch in `PosixSignals::init()`, to precede hotspot signal initialization. See my modified example: 

https://github.com/openjdk/jdk/commit/77826cf8253e5bda0ca7e694a2d100e4988d9e38

Then it works as planned, with SIGQUIT now being part of the normal hotspot VM signal installation.

**Comment by navyxliu** [comment]
I try it out. yeah, it works! One installation but it will remember early SIGBREAK.

Why do you insist to move jdk_misc_signal_init() before install_signal_handlers()? Is there any difference before and after it?

**Comment by caoman** [comment]
Moved `jdk_misc_signal_init()` earlier as Thomas suggested. Now that libjsig can prevent user overriding SIGBREAK later.
For Windows, `jdk_misc_signal_init()` seems appropriate as there's a `SetConsoleCtrlHandler()` above.

>  Why do you insist to move jdk_misc_signal_init() before install_signal_handlers()? Is there any difference before and after it?

Because SIGBREAK's UserHandler() uses the `pending_signals` and `sig_semaphore` variable initialized by `jdk_misc_signal_init()`.

File: src/hotspot/os/posix/signals_posix.cpp (hunk 4 above)
Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp]: The behavior has changed. Now that with libjsig, user must have -Xrs to install a custom SIGQUIT handler.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+ * @library /vmTestbase
+ *          /test/lib
+ * @run driver TestBreakSignalThreadDump
+ */
+
+/*
+ * @test id=with_jsig
+ * @bug 8292695
+ * @summary Check that Ctrl-\ causes HotSpot VM to print a full thread dump when signal chaining is used.
+ * @requires os.family == "linux" | os.family == "mac"
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:38]: Why not just !Windows? Aix, the BSDs, Solaris if this is downported should work too.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+public class TestBreakSignalThreadDump {
+
+    static class TestProcess {
+        static {
+            System.loadLibrary("ProcessUtils");
+        }
+
+        public static void main(String[] argv) throws Exception {
+            ProcessUtils.sendCtrlBreak();
+            System.out.println("Done!");
```
Comment (tstuefe) [inline test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:62]: Maybe don't finish right away but wait a second or two, for dispatcher thread on slow systems to catch up.

**Review by tstuefe — CHANGES_REQUESTED** [review]
Can you also add a test that checks that:

if VM is started with -Xrs, and we send a SIGQUIT, the VM shall crash with a core. The crash should look like a system crash, not an hs-err file crash, like this:

```
Quit (core dumped)
```

which assures us that no hotspot signal handler is installed and we take therefore the default action on SIGQUIT.

If this string is OS dependent and gives us headaches, the test should check at least that the process exit code is not 0.

And of course the VM should be started with tiny heap etc such to make sure the core file generated will be tiny.

File: src/hotspot/os/posix/signals_posix.cpp [hunk 7]
```diff
@@ -1310,11 +1304,24 @@ void install_signal_handlers() {
...
+    // The actual work for handling BREAK_SIGNAL is performed by the Signal
+    // Dispatcher thread, which is created and started at a much later point,
+    // see os::initialize_jdk_signal_support(). Any BREAK_SIGNAL received
+    // before the Signal Dispatcher thread is started is queued up via the
+    // pending_signals[BREAK_SIGNAL] counter, and will be processed by the
+    // Signal Dispatcher thread in a delayed fashion.
+    //
+    // Also note that HotSpot does NOT support signal chaining for BREAK_SIGNAL.
+    // Applications that require a custom BREAK_SIGNAL handler should run with
+    // -XX:+ReduceSignalUsage. Otherwise if libjsig is used together with
```
[the remark this answers is not in the document]
Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp:1319]: The SIGBREAK handler does not seem to honor `-XX:+AllowUserSignalHandlers`. It is still installed even with `-XX:+AllowUserSignalHandlers`.

Perhaps it can be addressed separately later? As commented in https://bugs.openjdk.org/browse/JDK-8293466, I haven't seen a real use case for +AllowUserSignalHandlers, so I'd vote for deprecating `AllowUserSignalHandlers` for removal in the future.

[the remark this answers is not in the document]
Comment (tstuefe) [reply src/hotspot/os/posix/signals_posix.cpp:1319]: Okay!

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java (hunk 5 above)
Comment (tstuefe) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:38]: Well, Solaris support was removed from JDK in 16, but older releases still support it, see: https://docs.oracle.com/en/java/javase/11/troubleshoot/handle-signals-and-exceptions.html#GUID-F012840D-8E5A-4B6D-8E49-0129A9EE46DB.

As  for AIX, the reason given for the exclusions in other tests is that we cannot run in the primordial thread on AIX. That is true but I'm not sure what this has to do with the signal tests, and I believe libjsig and the other stuff should just work on AIX. I have to dig in our older sources (this stems from the pre-OpenJDK times at SAP).

But to be consistent, I'd use !windows && !aix. AIX we can take care of later, but Solaris I would like to support here. Because of backporting, mainly.

File: src/hotspot/os/posix/signals_posix.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (caoman) [reply src/hotspot/os/posix/signals_posix.cpp:1319]: Thanks. Filed https://bugs.openjdk.org/browse/JDK-8293580.

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java (hunk 5 above)
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java]: Done. Thanks for the details!

File: test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java (hunk 6 above)
Comment (caoman) [reply test/hotspot/jtreg/runtime/Thread/TestBreakSignalThreadDump.java:64]: Done.

**Comment by caoman** [comment]
> Can you also add a test that checks that:
>
> if VM is started with -Xrs, and we send a SIGQUIT, the VM shall crash with a core. The crash should look like a system crash, not an hs-err file crash, like this:

I struggled to get this to work. I found out it is due to https://bugs.openjdk.org/browse/JDK-8234262, and ProcessBuilder changes the behavior of how SIGQUIT is handled in the subprocess. This was further complicated by how jtreg runs, as the workaround (passing -Xrs to the parent process) does not work through jtreg.

I uploaded the not-working test in here: https://github.com/caoman/jdk/commit/73e3a5f90985fcad9a383c704429cefb07011932

**Comment by tstuefe** [comment]
> > Can you also add a test that checks that:
> > if VM is started with -Xrs, and we send a SIGQUIT, the VM shall crash with a core. The crash should look like a system crash, not an hs-err file crash, like this:
> 
> I struggled to get this to work. I found out it is due to https://bugs.openjdk.org/browse/JDK-8234262, and ProcessBuilder changes how SIGQUIT is handled in the subprocess. This is further complicated by how jtreg runs the parent process. The workaround (passing -Xrs to the parent process) does not work through jtreg using "main/othervm -Xrs".
> 
> I uploaded the not-working test in here: [caoman@73e3a5f](https://github.com/caoman/jdk/commit/73e3a5f90985fcad9a383c704429cefb07011932)

Thank you for trying, and finding this out! I'm fine with postponing this test then to another RFE.

**Review by tstuefe — APPROVED** [review]
Looks good. Thank you for taking my suggestion!

..Thomas

**Review by navyxliu — APPROVED** [review]
the new revision still LGTM.

**Comment by caoman** [comment]
> Thank you for trying, and finding this out! I'm fine with postponing this test then to another RFE.

SG. Filed https://bugs.openjdk.org/browse/JDK-8293611.

**Comment by dholmes-ora** [comment]
Messing around with initialization order always makes me very nervous. While this may have solved the problem at hand it is far from clear to me that it has not introduced unexpected behaviour in other situations - situations that our testing is unlikely to expose.

**Comment by tstuefe** [comment]
> Messing around with initialization order always makes me very nervous. While this may have solved the problem at hand it is far from clear to me that it has not introduced unexpected behaviour in other situations - situations that our testing is unlikely to expose.

Hmm, we put a lot of work and time into this solution. While I acknowledge this is a difficult area, so are many others in hotspot which we are not shy of changing. And we did not change the coding on a whim - the proposed solution IMHO is better than the assortment of piled-on hotfixes we had before and dissolves quite a bit of complexity. That makes the code easier to understand and maintain in the long run. @caoman also provided a regression test, so that's covered too.

But if you have concrete misgivings about the proposed solution, let's discuss it.

**Comment by dholmes-ora** [comment]
@tstuefe IIUC the earlier proposal from @caoman  simply moved the setting of the BREAK_HANDLER to inside `jdk_misc_signal_init`. But you then had it moved instead to `install_signal_handlers`, which then required that `install_signal_handlers` get moved after `jdk_misc_signal_init` as the `UserHandler` depends on initialization that occurs there. I don't understand why that complex double-shuffle was needed/wanted instead of what Man proposed??

**Comment by tstuefe** [comment]
> @tstuefe IIUC the earlier proposal from @caoman simply moved the setting of the BREAK_HANDLER to inside `jdk_misc_signal_init`. But you then had it moved instead to `install_signal_handlers`, which then required that `install_signal_handlers` get moved after `jdk_misc_signal_init` as the `UserHandler` depends on initialization that occurs there. I don't understand why that complex double-shuffle was needed/wanted instead of what Man proposed??

See my comment at https://github.com/openjdk/jdk/pull/9955#issuecomment-1239643684. The original solution by @caoman , while it would have worked, was a bandaid atop of a bandaid (see also https://bugs.openjdk.org/browse/JDK-8279124). It was perpetuating a crooked solution, and the crookedness stems from the fact that SIGQUIT is semantically a VM signal, owned by the VM, but initialized outside the libjsig-VM-signal-initialization window. The solution proposed by me makes SIGQUIT initialize like any other VM signal, therefore we can remove the SIGQUIT-related special handling and hopefully don't need further patches in this area.

**Comment by dholmes-ora** [comment]
I don't really agree with the rationale, but the fix is correct, and the reordering seems harmless enough on further inspection.

To me this is a JDK serviceability signal, both for dynamic attach and thread dumps, as it is the JDK serving "signal thread" that handles it. Most of the problems in this area have stemmed from initialization issues due to "too early" sending of SIGQUIT to initiate an attach attempt.

**Comment by caoman** [comment]
> SIGQUIT is semantically a VM signal, owned by the VM, but initialized outside the libjsig-VM-signal-initialization window. The solution proposed by me makes SIGQUIT initialize like any other VM signal, ...

Yes, this is main reason we installed BREAK_SIGNAL handler inside `install_signal_handlers` and reordered `jdk_misc_signal_init` and `install_signal_handlers`.

There was a concern that with libjsig present and without `-Xrs`, user could still install a custom BREAK_SIGNAL handler that overrides JVM's handler, see https://github.com/openjdk/jdk/pull/9955#discussion_r956623350. This change avoids this problem.
